modules/fileout_netcdf/d4_tools.cc

```diff
@@ -0,0 +1,193 @@
+        case libdap::dods_grid_c:
+        case libdap::dods_sequence_c: {
+            auto *svar = dynamic_cast<Constructor*>(var);
+            bool is_dap4 = false;
+            for (auto bvar = svar->var_begin(), evar = svar->var_end(); bvar != evar; ++bvar){
+
+                if (d4_tools::is_dap4_projected(*bvar, inv)){
+                    is_dap4 = true;
+                }
+            }
+            if(is_dap4) return true;
+        }
+        default:
+            break;
+    }
+    // Projected variable might not be DAP4, but what about the attributes?
+    if(has_dap4_types(var->attributes())){
+        // It has DAP4 Attributes so that variable is a de-facto dap4 variable.
+        inv.push_back(var);
+        return true;
+    }
+    return false;
+}
+
+#if 1
+bool d4_tools::is_dap4_projected(DDS *_dds, vector<BaseType *> &projected_dap4_variable_inventory) {
+    bool has_dap4_var = false;
+    for (auto btp = _dds->var_begin(), ve = _dds->var_end(); btp != ve; ++btp) {
+
+        BaseType *var = *btp;
+        if(var->send_p()){
+            if (is_dap4_projected(var, projected_dap4_variable_inventory)) {
+                has_dap4_var = true;
+            }
+        }
+    }
+    return has_dap4_var;
+}
+#endif
+
+#if 1
+bool d4_tools::is_dap4_projected(DMR *_dmr, vector<BaseType *> &projected_dap4_variable_inventory) {
+
+    bool has_dap4_var = false;
+
+    has_dap4_var = is_dap4_projected( _dmr->root(), projected_dap4_variable_inventory);
+
+    return has_dap4_var;
+}
+#endif
+
+#if 1
+bool d4_tools::is_dap4_projected(D4Group *group, vector<BaseType *> &projected_dap4_variable_inventory){
+    bool has_dap4_var = false;
+    // variables in the group
+
+    // check all variables in the root group
+    for (auto btp = group->var_begin(), ve = group->var_end(); btp != ve; ++btp) {
+        BaseType *var = *btp;
+
+        if(var->send_p() && var->type() != libdap::dods_group_c){ //dods_group_c <- d4Groups
+            if (is_dap4_projected(var, projected_dap4_variable_inventory)) {
+                has_dap4_var = true;
+            }
+        }
+        else if (var->send_p() && var->type() == libdap::dods_group_c) {
+            try {
+                D4Group *d4g = dynamic_cast<D4Group *>(var);
+                if (is_dap4_projected(d4g, projected_dap4_variable_inventory)) {
+                    has_dap4_var = true;
+                }
+            }
+            catch (exception e){
+                throw BESInternalError("Exception: " + string(e.what()), __FILE__, __LINE__);
+            }
+        }
+    }
+
+    //check if group has any subgroups and check them for dap4 projected variables
+    for (auto gb = group->grp_begin(), ge = group->grp_end(); gb != ge; ++gb){
+        try {
+            D4Group *d4g = *gb;
+            if (is_dap4_projected(d4g, projected_dap4_variable_inventory)) {
+                has_dap4_var = true;
+            }
+        }
+        catch (exception e){
```

Comment:
Same as above.

Reply:
One of the flagged 'code smells' is to change the '#define d4_tools_h_ 1' line to a constexpr. Ignore that suggestion. The other suggestions are actually pretty good.